parcels/codegenerator.py

```diff
@@ -922,48 +922,81 @@ def generate(self, funcname, field_args, const_args, kernel_ast, c_include):
         if c_include:
             ccode += [c_include]
 
-        # Insert kernel code
+        # ==== Insert kernel code ==== #
         ccode += [str(kernel_ast)]
 
         # Generate outer loop for repeated kernel invocation
         args = [c.Value("int", "num_particles"),
                 c.Pointer(c.Value(self.ptype.name, "particles")),
-                c.Value("double", "endtime"), c.Value("float", "dt")]
+                c.Value("double", "endtime"),
+                c.Value("double", "dt")
+                ]
         for field, _ in field_args.items():
             args += [c.Pointer(c.Value("CField", "%s" % field))]
         for const, _ in const_args.items():
-            args += [c.Value("float", const)]
+            args += [c.Value("double", const)]
         fargs_str = ", ".join(['particles[p].time'] + list(field_args.keys())
                               + list(const_args.keys()))
-        # Inner loop nest for forward runs
+        # ==== statement clusters use to compose 'body' variable and variables 'time_loop' and 'part_loop' ==== ##
         sign_dt = c.Assign("sign_dt", "dt > 0 ? 1 : -1")
         particle_backup = c.Statement("%s particle_backup" % self.ptype.name)
-        sign_end_part = c.Assign("sign_end_part", "endtime - particles[p].time > 0 ? 1 : -1")
-        dt_pos = c.Assign("__dt", "fmin(fabs(particles[p].dt), fabs(endtime - particles[p].time))")
+        sign_end_part = c.Assign("sign_end_part", "(endtime - particles[p].time) > 0 ? 1 : -1")
+        reset_res_state = c.Assign("res", "particles[p].state")
+        update_state = c.Assign("particles[p].state", "res")
+        update_pdt = c.If("_next_dt_set == 1",
+                          c.Block([c.Assign("_next_dt_set", "0"), c.Assign("particles[p].dt", "_next_dt")]))
+
+        dt_pos = c.Assign("__dt", "fmin(fabs(particles[p].dt), fabs(endtime - particles[p].time))")                   # original
+        # dt_pos = c.Assign("__dt", "fmin(fabs(particles[p].dt), (endtime - particles[p].time) * sign_dt)")           # reint
+        # dt_pos = c.Assign("__dt", "fmax(fmin(fabs(particles[p].dt), (endtime - particles[p].time) * sign_dt), 0)")  # christian
```

Comment:
Remove these commented lines?

Reply:
yes, I can do that - left it in until getting the final OK from Reint